development/webpack/utils/cli.ts

```diff
@@ -0,0 +1,391 @@
+ * build is a production build or not.
+ *
+ * @param argv
+ * @param opts
+ * @returns `true` if this is a production build, otherwise `false`
+ */
+function preParse(
+  argv: string[],
+  opts: typeof prerequisites,
+): { env: 'production' | 'development' } {
+  const options: { [k: string]: { [k: string]: unknown } } = {
+    configuration: {
+      envPrefix: ENV_PREFIX,
+    },
+  };
+  // convert the `opts` object into a format that `yargs-parser` can understand
+  for (const [arg, val] of Object.entries(opts)) {
+    for (const [key, valEntry] of Object.entries(val)) {
+      if (!options[key]) {
+        options[key] = {};
+      }
+      options[key][arg] = valEntry;
+    }
+  }
+
+  const { env } = parser(argv, options);
+  return { env };
+}
+
+/**
+ * Type representing the parsed arguments
+ */
+export type Args = ReturnType<typeof parseArgv>['args'];
+export type Features = ReturnType<typeof parseArgv>['features'];
+
+/**
+ * Parses an array of command line arguments into a structured format.
+ *
+ * @param argv - An array of command line arguments, excluding the program
+ * executable and file name. Typically used as
+ * `parseArgv(process.argv.slice(2))`.
+ * @param buildConfig - The build config.
+ * @param buildConfig.buildTypes - The build types.
+ * @param buildConfig.features - The features.
+ * @returns An object representing the parsed arguments.
+ */
+export function parseArgv(
+  argv: string[],
+  { buildTypes, features }: BuildConfig,
+) {
+  const allBuildTypeNames = Object.keys(buildTypes);
+  const allFeatureNames = Object.keys(features);
+
+  // args like `production` may change our CLI defaults, so we pre-parse them
+  const preconditions = preParse(argv, prerequisites);
+  const options = getOptions(preconditions, allBuildTypeNames, allFeatureNames);
+  const args = getCli(options, 'yarn webpack').parseSync(argv);
+  // the properties `$0` and `_` are added by yargs, but we don't need them. We
+  // transform `add` and `omit`, so we also remove them from the config object.
+  const { $0, _, addFeature: add, omitFeature: omit, ...config } = args;
+
+  // set up feature flags
+  const active = new Set<string>();
+  const defaultFeaturesForBuildType = buildTypes[config.type].features ?? [];
+  const setActive = (f: string) => omit.includes(f) || active.add(f);
+  [defaultFeaturesForBuildType, add].forEach((feat) => feat.forEach(setActive));
+
+  const ignore = new Set(['$0', 'conf', 'progress', 'stats', 'watch']);
+  const cacheKey = Object.entries(args)
+    .filter(([key]) => key.length > 1 && !ignore.has(key) && !key.includes('-'))
+    .sort(([x], [y]) => x.localeCompare(y));
+  return {
+    // narrow the `config` type to only the options we're returning
+    args: config as { [key in OptionsKeys]: (typeof config)[key] },
+    cacheKey: JSON.stringify(cacheKey),
+    features: {
+      active,
+      all: new Set(allFeatureNames),
+    },
+  };
+}
+
+/**
+ * Gets a yargs instance for parsing CLI arguments.
```

Comment:
not a blocker, but maybe we could have a config.json file example somewhere, and we could also add an entry to gitignore, for a seamless dev experience?
It could also be added in a future PR if we want it

Reply:
What do you mean by "config.json" here?

Reply:
oh, I meant for when we pass the flag `--config`, so we have a template file to reference, and quicker setup, but not really needed

Reply:
Ah yeah, good idea. I've added some usage instructions for using `--config` as well as using environment variables to set options in the read me. If using `--config` catches on we should look into standardizing an example `--config` file (and adding it to .gitignore) like we do now for `.metamaskrc`. I don't think we should do it yet, as I'm not too sure the feature will be used all too often, if ever.